Allow user to opt-out and delete their own information if they do not wish to have an account with gitpay

yan-shcherbakov opened the pull request:
Hi,

I have been trying to get a hold of someone at gitpay to have my account removed so that it can also be purged from Google search.

My initial contact was on Dec. 21st, 2016 and I have not heard back since despite two additional attempts.

Gitpay seems to create an "unactivated" account for anyone that happens to click on the URI and then upon authenticating via github oath, it then creates an account but does not allow you in anyway to manage your account or even delete your own account if you do not wish to have a gitpay account.

I imagine that many people like myself have attempted to "activate" their account simply for the purpose of deleting their own account. Therefore now they can without having to contact anyone.

So I figured that I can simply create a pull request for an easy way to simply delete all information from the database that relates to their particular account using their logged in session information. 

This is extremely useful for many as your website states: "Account management and more features coming soon." and it is not easy to get in contact with anyone at gitpay.

I also noticed that in the database, there's a flag `deleted` which is either 0 or 1. This hints to me that if an account is deleted, the information may still be there. I discourage you (gitpay) from doing this since if an account is considered deleted, then it must be purged from the database. You should generally not continue to store one's information after they have deleted their account.

I understand that this may be useful in case your system re-creates an account for them later, but this should probably be handled by another table of previously deleted github uris.

Emails, names and so forth should be purged upon delete.

This resolves issue #5 .

Commit:
Adding user opt-out functionality.

Commit:
Adding the files to remove the users.

yan-shcherbakov commented:
@melvincarvalho @Erkan-Yilmaz please review.

ahmadnassri commented:
I would like my name and info be deleted as well, I never authorized gitpay to use my account info!!!

yan-shcherbakov commented:
Bump.

melvincarvalho commented:
@yan-shcherbakov thanks for raising an issue and pull request

I am currently on vacation and will review this when I return in the new year

May I request that you please do not initiate. unsolicited mails to members of the github team over the christmas holidays, especially.  I think you'll agree 7 mails in as many days is rather aggressive, for a time traditionally set aside for family.

Let me just point out that the information on the site is of two forms.  One is the public information that you consented to make public and to be replicated / forked ie repository information, and the followers and followings of your social coding graph.  If you show up in the site it's because someone followed your work publicly or you followed someone's work, leading to my work.  Inevitably there will be issues with freshness and cache invalidation, associated with that data, and that will simply be a function of how much resource is employed to keep data fresh.  The other data is opt-in to allow you to receive donations for open source work.  This is content created by the site, and it should be possible to delete that.

Thanks for the pull request, that's helpful, I'll review this in the new year.

yan-shcherbakov commented:
@melvincarvalho I think you will agree that creating accounts and using information without their consent is unacceptable regardless of the time of the year. It is your responsibility to remove information that violates a person's privacy in a timely manner.

Also, I don't agree that my communication has been aggressive as I've only sent you 4 messages to your personal emails since Dec. 21st. I'm not sure where you got the other 3 mails from.

Even though I consented to have my information public, I did not consent to the information being available on a third party service when it was added. Third party service may use public information, but must address take down requests.

melvincarvalho commented:
> I'm not sure where you got the other 3 mails from.

@yan-shcherbakov you may or may not be aware creating an issue on here triggers an email to be sent to those watching the repository

By signing up to github social coding, and setting your profile to public, you consented to make your repository information public and to consent to allowing it to be forked.  So anything published is part of the public domain, and something that you opted in to.  The only private information is your donation address and that is strictly opt-in.  The issue you raise is related to freshness of information.  

Thanks again, I will look at your pull request in the new year.

yan-shcherbakov commented:
@melvincarvalho I have no issue if you choose to address my PR in the new year.

However, I urge you to please remove my information from your website as requested on Dec. 21st. This is not a huge request, and so far I have been constructive in attempting to resolve this matter.

I do not consent in having this information being available on your website.

yan-shcherbakov commented:
The information that I provided to github is copyright and in this case belongs to github. I have given github permission to use my information for forking my repos and so forth. You, as a third party, you do not have the license to use my information especially my personal name.

Forking remains under the github domain.

Also, under DMCA you have the obligation to address take down request in a timely manner. While there's no time limit to addressing the concerns, you should not hesitate.

http://www.dmlp.org/legal-guide/responding-dmca-takedown-notice-targeting-your-content

minimaxir commented:
> By signing up to github social coding, and setting your profile to public, you consented to make your repository information public and to consent to allowing it to be forked. So anything published is part of the public domain, and something that you opted in to.

Technically incorrect. The data is *owned* by GitHub, and subject to their [Terms of Service](https://help.github.com/articles/github-terms-of-service/) for any method in retrieving it (e.g. the API).

yan-shcherbakov commented:
https://help.github.com/articles/github-terms-of-service/#f-copyright-and-content-ownership

Github TOS states:

> We claim no intellectual property rights over the material you provide to the Service. Your profile and materials uploaded remain yours. However, by setting your pages to be viewed publicly, you agree to allow others to view your Content. By setting your repositories to be viewed publicly, you agree to allow others to view and fork your repositories.

Note:

> you agree to allow others to view your Content.

So, I allowed my information to be viewed, yes. Not re-published on a third party website.

I allowed my work to be forked, yes... but that remains within github if my work is forked.

ugexe commented:
@melvincarvalho How are emails from watching a github repo unsolicited? A better example of the meaning of unsolicited would be creating shadow accounts to represent people that have not requested it.

You did this yourself, so you have only yourself to blame if this affects your family time.

sehrope commented on functions.php:
```diff
@@ -100,6 +100,18 @@ function getUser($nick, $conn = null, $client = null) {
 
+function deleteUser($nick, $conn = null, $client = null) {
+  if ($conn) {
+    $sql = "delete from users where login = '$nick'; ";
```
I'm not intimately familiar with PHP but I'm pretty sure this is vulnerable to SQL injection as it's blindly concatenating the user supplied nick. It should be escaped or (*much better*) be supplied as a bind variable (i.e. parameterized statement).

yan-shcherbakov replied:
I agree, this is not the most optimal solution. I should change this to a `sprintf` statement, but the after this: `$stmt = $conn->prepare($sql);` should take care of this issue I believe.

sehrope replied:
`sprintf(...)` is not appropriate for escaping SQL literals and identifiers as it doesn't handle SQL quoting/escaping semantics. The preparing of the concatenated SQL does nothing to make it safer.

As I believe this PHP app is using PDO, you should be using `$stmt.bindParam(...)`: http://php.net/manual/en/pdostatement.bindparam.php

yan-shcherbakov replied:
Committed an edit to this.

sehrope replied:
There are two SQL commands being executed but you only parametrized one of them. The other on line 106/107 still has the same issue as before.

yan-shcherbakov replied:
Thanks for pointing that out.

yan-shcherbakov replied:
Now the 2nd on is parametrized. Forgot that I had two statements.

Commit:
Updating delete function with prepared statement.

melvincarvalho commented:
> How are emails from watching a github repo unsolicited? A better example of the meaning of unsolicited would be creating shadow accounts to represent people that have not requested it.

Sorry, just for clarification.  

yan-shcherbakov also sent multiple unsolicited emails to one of my email accounts (which is not my primary email account, so I had to look back).  I did get a bunch of email over the Christmas weekend, which was disruptive.  

Watching of the repo was less disruptive, as I can turn it off.  Raising an issue the day after christmas and bumping it one day later is rather aggressive, by anyone's standards.

Thanks for reviewing the pull request.

franciscop commented on user.php:
```diff
@@ -255,8 +256,9 @@
-          echo "<h4>Congratulations, you have logged in.  You have earnt one reputation point.  Account management and more <a target=\"_blank\" href=\"https://melvincarvalho.gitbooks.io/gitpay/content/\">features</a> coming soon.<h4><hr>";
+        <?php
+        if( isset($_SESSION['loggedin']) && $_SESSION['loggedin'] === 'true' && $_SESSION['login'] === $nick  ) {
+          echo "<h4>Congratulations, you have logged in.  You have earnt one reputation point.  Account management and more <a target=\"_blank\" href=\"https://melvincarvalho.gitbooks.io/gitpay/content/\">features</a> coming soon.<h4><h4>If you wish to opt-out of gitpay, click here: <a href='/deleteuser.php'>Remove my account</a></h4><hr>";
```
I would totally add here a confirmation step (non-revokable action) and an URL that doesn't leak the extension:

```html
<a class="delete" href="/deleteuser">...
```

Confirmation:

```js
document.querySelector('a.delete').addEventListener('click', function(e) {
  var result = confirm("Are you sure?");
  if (!result) e.preventDefault();
});
```

yan-shcherbakov commented:
@melvincarvalho I only sent 4 emails and one of them was on the 26th. I don't see how that is disruptive.

Also, those emails don't fall under the definition of unsolicited emails. Those are take down requests, and you have your email listed as an administrative contact under your whois. You don't provide an alternate contact on gitpay.org.

On further emails, I have simply been following up on my request. I will gladly post my emails to you publicly to show that I have been respectful towards you in my requests.

If you had responded to my initial email on Dec. 21st simply saying "I'm on vacation will attend to your concern after the New Year" at least I would have expected to hear back from you in the New Year and refrained from bothering you during this time.

Commit:
Forgot to parameterize additional statement. My bad.

melvincarvalho commented:
merged

yan-shcherbakov commented:
Thank you! @melvincarvalho
